h/static/scripts/controllers/share-widget-controller.js

```diff
@@ -0,0 +1,212 @@
+        this.hide();
+      }
+    };
+
+    window.document.body.addEventListener('click', this._handler);
+  }
+
+  /**
+   * @typedef {Object} ConfigOptions
+   * @property {String} url - the url we are enabling to be shared
+   * @property {Bool} [private] - is the card only visible to this user
+   * @property {Bool} [group] - is the card posted in a group scope
+   */
+
+  /**
+   * Update the template based on the config variables passed in
+   *
+   * @param  {ConfigOptions} config The details we need to apply update our template
+   *   with the correct information per card.
+   */
+  _renderWidgetTemplate(config) {
+
+    // copy to clipboard input
+    this._widget.querySelector(CLIPBOARD_INPUT_SELECTOR).value = config.url;
+
+    // social links
+    Array.from(this._widget.querySelectorAll(TARGET_HREF_SELECTOR)).forEach((target) => {
+      target.href = target.getAttribute(TARGET_HREF_ATTR).replace('{href}', encodeURI(config.url));
+    });
+
+    // scope access dialog
+    const privateMessage = this._widget.querySelector(PRIVATE_MSG_SELECTOR);
+    const groupMessage = this._widget.querySelector(GROUP_MSG_SELECTOR);
+
+    privateMessage.style.display = 'none';
+    groupMessage.style.display = 'none';
+
+    if (config.private) {
+      privateMessage.style.display = 'block';
+    } else if (config.group) {
+      groupMessage.style.display = 'block';
+    }
+  }
+
+
+  /**
+   * Given a node, update the template, repostion the widget properly,
+   *  and make it visible.
+   *
+   * @param  {HTMLElement} node The trigger node that we will place the widget
+   *   next to.
+   * @param  {ConfigOptions} config Passed through to rendering/interpolation
+   */
+  showForNode(node, config) {
+
+    if (!node || !config) {
+      return;
+    }
+
+    this._renderWidgetTemplate(config);
+
+    // offsets affecting height need to be updated after variable replacement
+    const widgetOffsets =  _getOffset(this._widget);
+    const nodeOffset = _getOffset(node);
+
+    this._widget.style.top = (nodeOffset.top - widgetOffsets.height - ARROW_PADDING_BOTTOM) + 'px';
+    this._widget.style.left = ((ARROW_PADDING_RIGHT + nodeOffset.left + (nodeOffset.width / 2)) - widgetOffsets.width) + 'px';
+
+    this._container.style.visibility = 'visible';
+
+    this._widgetVisible = true;
+  }
+
+  hide() {
+    this._container.style.visibility = 'hidden';
+    this._widgetVisible = false;
+  }
+
+
+  /**
+   * Utility to clean up the listeners applied. Otherwise the subsequent
+   * constructor will reset all other state. Primary use is meant for testing cleanup
+   */
+  detach() {
+    window.document.body.removeEventListener('click', this._handler);
+  }
+}
+
+
+// Note, this seemlingly meaningless class is here so that we can map
+// our controller paradigm to the invokaction of the appropriate events. And leave
+// the ShareWidget compontent itself abstracted from this paradigm.
+class ShareWidgetController extends Controller {
```

Comment:
> And leave the ShareWidget compontent itself abstracted from this paradigm.

I don't think we need this really, the `Controller` class is very little code. We could just include it as part of any package of reusable code that is shared with the client or other projects. In the interests of getting this shipped - let's leave that discussion for later though.

What would be useful here is a JSDoc comment explaining what a `ShareWidgetController` is.